gensim/models/ldamodel.py

```diff
@@ -1020,28 +1022,41 @@ def diff(self, other, distance="kullback_leibler", num_words=100, n_ann_terms=10
         if distance == "jaccard":
             d1, d2 = fst_topics, snd_topics
 
-        z = np.zeros((t1_size, t2_size))
-        for topic1 in range(t1_size):
-            for topic2 in range(t2_size):
-                z[topic1][topic2] = distance_func(d1[topic1], d2[topic2])
-
-        if normed:
-            if np.abs(np.max(z)) > 1e-8:
-                z /= np.max(z)
-
-        annotation = [[None] * t1_size for _ in range(t2_size)]
+        if diagonal:
+            assert t1_size == t2_size, "Both input models should have same no. of topics, as the diagonal will only be valid in a square matrix"
+            # initialize z and annotation array
+            z = np.zeros(t1_size)
+            if annotation:
+                diff_terms = np.zeros(t1_size, dtype=list)
+        else:
+            # initialize z and annotation matrix
+            z = np.zeros((t1_size, t2_size))
+            if annotation:
+                diff_terms = np.zeros((t1_size, t2_size), dtype=list)
+
+        # iterate over each cell in the initialized z and annotation
+        for topic in np.ndindex(z.shape):
+            topic1 = topic[0]
+            if diagonal:
+                topic2 = topic1
+            else:
+                topic2 = topic[1]
 
-        for topic1 in range(t1_size):
-            for topic2 in range(t2_size):
+            z[topic] = distance_func(d1[topic1], d2[topic2])
+            if annotation:
                 pos_tokens = fst_topics[topic1] & snd_topics[topic2]
                 neg_tokens = fst_topics[topic1].symmetric_difference(snd_topics[topic2])
 
                 pos_tokens = sample(pos_tokens, min(len(pos_tokens), n_ann_terms))
```

Comment:
As I remember, you want to remove `sample` for another PR, you already do it in different PR?